A74-xds-config-tears.md

```diff
@@ -0,0 +1,311 @@
+  // Called when a new xDS config is available.
+  virtual void OnUpdate(XdsConfig config) = 0;
+
+  // These methods are invoked when there is an error or
+  // does-not-exist on LDS or RDS only.  The context will be a
+  // human-readable string indicating the scope in which the error
+  // occurred (e.g., the resource type and name).
+  virtual void OnError(absl::string_view context, absl::Status status) = 0;
+  virtual void OnResourceDoesNotExist(std::string context) = 0;
+};
+```
+
+The xds resolver will implement this watcher interface and pass the
+watcher object to the XdsDependencyManager at instantiation time.
+
+The updates delivered to this watcher will contain a complete xDS
+configuration, including all necessary resources.  The xDS configuration
+will look something like this (C++ syntax, various implementation
+details elided):
+
+```c++
+struct XdsConfig {
+  // Listener resource.
+  XdsListenerResource listener;
+  // RouteConfig resource.  Will be populated even if RouteConfig is
+  // inlined into the Listener resource.
+  XdsRouteConfigResource route_config;
+  // Virtual host.  Points into route_config.  Will always be non-null.
+  dsRouteConfigResource::VirtualHost* virtual_host;
+
+  // Cluster map.  A cluster will have a non-OK status if either
+  // (a) there was an error and we did not already have a valid
+  // resource or (b) the resource does not exist.
+  // For aggregate clusters, the aggregate cluster will be first in
+  // the list, followed by the list of leaf clusters.  For other types
+  // of clusters, the list will have only one element.
+  struct ClusterConfig {
+    // Cluster name and resource.
+    std::string cluster_name;
+    XdsClusterResource cluster;
+    // Endpoint info.  If there was an error, endpoints will be null
+    // and resolution_note will be set.
+    XdsEndpointResource endpoints;
+    std::string resolution_note;
+  };
+  std::map<std::string, absl::StatusOr<std::vector<ClusterConfig>>> clusters;
+};
+```
+
+Note that the watcher's OnError() and OnResourceDoesNotExist()
+methods will be invoked only for problems with the listener and route
+configuration resources.  For cluster-specific errors, we need to
+retain the existing behavior, which is that it causes a problem for the
+individual cluster but does not affect traffic sent to any other cluster.
+Therefore, when the XdsDependencyManager sees an error on a cluster for
+which it did not previously have data, or if it sees a does-not-exist
+notification on a cluster, it will report that by returning a non-OK
+status for the cluster in the XdsConfig.  Similarly, if there is a problem
+obtaining endpoint data from EDS or DNS, the XdsDependencyManager will
+report that problem via the cluster's resolution_note field.
+
+The xds resolver will pass the returned XdsConfig to the LB policies via
+an attribute, so that the LB policies can use the cluster data.  Note
+that we cannot simply pass the data via the cds LB policy's JSON config,
```

Comment:
In the dynamic cluster case, we don't have a way to pass the config down as JSON, because the config of the child is coming from the ClusterSpecifierPlugin policy, not from the xds resolver.

I've added some wording to try to clarify this.